### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by otaviobp**
Introducing Unit tests in swupd.

### CODE REVIEW

File: src/helpers.c [hunk 1]
```diff
@@ -585,68 +585,6 @@ enum swupd_code swupd_init(void)
...
-int strtoi_err(const char *str, int *value)
-{
-	char *endptr;
-	int err = strtoi_err_endptr(str, &endptr, value);
-
-	if (err) {
-		return err;
-	}
-
-	if (*endptr != '\0' && !isspace(*endptr)) {
```
Comment (jwakre) [inline src/helpers.c]: Can we remove #include <ctype.h> now? It was added to support isspace(*endptr) which is removed.

Comment (otaviobp) [reply src/helpers.c]: I removed now ctype.h and types.h includes from helpers.c
Thanks.

File: src/lib/strings.h [hunk 2]
```diff
@@ -23,6 +23,34 @@ void free_string(char **s);
 /* Join strings from string list separated by the separator */
 char *string_join(const char *separator, struct list *string);
 
+/*
+ * strtoi_err: Safely convert and string to integer avoiding overflows
```
Comment (castulo) [inline src/lib/strings.h]: Isn't this suppose to say "Safely convert a~~nd~~ string to integer avoiding overflows."

File: src/lib/strings.h [hunk 3]
```diff
@@ -23,6 +23,34 @@ void free_string(char **s);
...
+ *
+ * This function returns negative error codes based on the errno table:
+ * -ERANGE is returned when the string is out of range for int value
+ * -EINVAL is returned when the string isn't a valid number or has any invalid
+ * trailing character.
+ */
+int strtoi_err(const char *str, int *value);
+
+/*
+ * strtoi_err: Safely convert and string to integer avoiding overflows.
```
Comment (castulo) [inline src/lib/strings.h]: wrong function name in the comment strtoi_err -> strtoi_err_endptr (and same typo as above)

File: Makefile.am [hunk 4]
```diff
@@ -261,13 +262,19 @@ BATS = \
 	test/functional/verify/verify-picky-whitelist.bats \
 	test/functional/verify/verify-skip-scripts.bats
 
+UNIT_TESTS = \
+	test/unit/test_strings.test
```
Comment (castulo) [inline Makefile.am]: This is not necessary for this PR, but I would really like a way of discovering these in the future. It is something I've been meaning to do for functional tests too so we don't have to add/remove every test to/from the list every time.

File: test/unit/test_strings.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+	// test EINVAL errors
+	char *einval_errors[] = { "string", "123.456", "123a", "a123a", "456b123", "0x12345", "4f"};
+
+	for (i = 0; i < sizeof(einval_errors) / sizeof(char *); i++) {
+		err = strtoi_err(einval_errors[i], &value);
+		check(err == -EINVAL);
+	}
+}
+
+static void test_strtoi_enderr()
```
Comment (castulo) [inline test/unit/test_strings.c]: shouldn't this be _test_strtoi_err_endptr_ instead of _test_strtoi_enderr_?

File: Makefile.am (hunk 4 above)
Comment (otaviobp) [reply Makefile.am]: +1 for this too, but I don't know how to do it. haha

**Comment by otaviobp** [comment]
fixed

**Pull request merged**